firmware-management/swupdate-handlers/arm-handler-common.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#include <mntent.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <sys/stat.h>
+#include "util.h"
+#include <arm-handler-common.h>
+
+char *malloc_or_abort(const size_t size)
+{
+    char *dst = malloc(sizeof dst * size);
+    if (!dst)
+    {
+        ERROR("%s", "Failed to allocate memory");
+        abort();
+    }
+    return dst;
+}
+
+void str_delete(char *str)
+{
+    free(str);
+}
+
+char *str_copy_to_new(const char *const src)
+{
+    const size_t dst_size = strlen(src) + 1;
+    char *dst = malloc_or_abort(dst_size);
+    strncpy(dst, src, dst_size);
+
+    if (dst[strlen(src)] != '\0')
+        dst[strlen(src)] = '\0';
+
+    return dst;
+}
+
+int str_endswith(const char *const substr, const char *const fullstr)
+{
+    const size_t substr_len = strlen(substr);
+    const size_t fullstr_len = strlen(fullstr);
+    const int endlen = fullstr_len - substr_len;
```

Comment:
I don't think this will work as intended when `fullstr_len` is less than `substr_len`: both variables are `size_t`s so integer promotion probably won't be applied so the type of the expression `fullstr_len - substr_len` will probably be `size_t` and the subtraction will use modular arithmetic.... Compiler explorer (with `-x c`) says that
```C
int main(){
    size_t a = 10;
    size_t b = 12;
    int c = a - b;
    printf("%d\n", c);
    return 0;
}
```
prints `-2` though. Switching to C++ and adding `printf("%s\n", typeid(decltype(a - b)).name());` prints `m`, which means `size_type`, so the type of the expression `a - b` is definitely unsigned. My guess is that `a - b` is a big unsigned integer that overflows the `int` (which is undefined behaviour) but the arithmetic just works out in the end. The rules are slightly different for C++ but I suspect the result applies to C too.

I suggest doing:
```C
if (substr_len > fullstr_len) return 1;
const int endlen = fullstr_len - substr_len;
// No need for the if (endlen < 0) check: we know that substr_len <= fullstr_len so endlen must be >= 0
```

Reply:
Okay, I'll change it in the next commit.